internal/batches/service/service.go

```diff
@@ -155,6 +155,110 @@ func (svc *Service) CreateChangesetSpec(ctx context.Context, spec *batcheslib.Ch
 	return graphql.ChangesetSpecID(result.CreateChangesetSpec.ID), nil
 }
 
+const resolveWorkspacesForBatchSpecQuery = `
+query ResolveWorkspacesForBatchSpec($spec: String!) {
+    resolveWorkspacesForBatchSpec(batchSpec: $spec) {
+		onlyFetchWorkspace
+		ignored
+		unsupported
+		repository {
+			id
+			name
+			url
+			externalRepository { serviceType }
+			defaultBranch {
+				name
+				target { oid }
+			}
+		}
+		branch {
+			name
+			target {
+				oid
+			}
+		}
+		path
+		searchResultPaths
+    }
+}
+`
+
+func (svc *Service) ResolveWorkspacesForBatchSpec(ctx context.Context, spec *batcheslib.BatchSpec, allowUnsupported, allowIgnored bool) ([]RepoWorkspace, []*graphql.Repository, error) {
+	raw, err := json.Marshal(spec)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "marshalling changeset spec JSON")
+	}
+
+	var result struct {
+		ResolveWorkspacesForBatchSpec []struct {
+			OnlyFetchWorkspace bool
+			Ignored            bool
+			Unsupported        bool
+			Repository         *graphql.Repository
+			Branch             *graphql.Branch
+			Path               string
+			SearchResultPaths  []string
+		}
+	}
+	if ok, err := svc.newRequest(resolveWorkspacesForBatchSpecQuery, map[string]interface{}{
+		"spec": string(raw),
+	}).Do(ctx, &result); err != nil || !ok {
+		return nil, nil, err
+	}
+
+	unsupported := batches.UnsupportedRepoSet{}
+	ignored := batches.IgnoredRepoSet{}
+
+	repos := make([]*graphql.Repository, 0, len(result.ResolveWorkspacesForBatchSpec))
+	seenRepos := make(map[string]struct{})
+	workspaces := make([]RepoWorkspace, 0, len(result.ResolveWorkspacesForBatchSpec))
+	for _, w := range result.ResolveWorkspacesForBatchSpec {
+		fileMatches := make(map[string]bool)
+		for _, path := range w.SearchResultPaths {
+			fileMatches[path] = true
+		}
+
+		workspace := RepoWorkspace{
+			Repo: &graphql.Repository{
+				ID:                 w.Repository.ID,
+				Name:               w.Repository.Name,
+				URL:                w.Repository.URL,
+				FileMatches:        fileMatches,
+				ExternalRepository: w.Repository.ExternalRepository,
+				DefaultBranch:      w.Repository.DefaultBranch,
+				Commit:             w.Branch.Target,
+				Branch:             *w.Branch,
+			},
+			Path:               w.Path,
+			OnlyFetchWorkspace: w.OnlyFetchWorkspace,
+		}
+
+		// Collect the repo, if not seen yet.
+		if _, ok := seenRepos[workspace.Repo.ID]; !ok {
+			seenRepos[workspace.Repo.ID] = struct{}{}
+			repos = append(repos, workspace.Repo)
+		}
+
+		if !allowIgnored && w.Ignored {
+			ignored.Append(workspace.Repo)
+		} else if !allowUnsupported && w.Unsupported {
+			unsupported.Append(workspace.Repo)
+		} else {
+			workspaces = append(workspaces, workspace)
+		}
+	}
+
+	if unsupported.HasUnsupported() {
+		return workspaces, repos, unsupported
+	}
+
+	if ignored.HasIgnored() {
+		return workspaces, repos, ignored
+	}
```

Comment:
This just mimics what we currently do here. This deserves a refactor, too, but not in this PR :)